src/vanilla.ts

```diff
@@ -43,240 +15,296 @@ type Op =
+  },
 
-export function subscribe<T extends object>(
-  proxyObject: T,
-  callback: (ops: Op[]) => void,
-  notifyInSync?: boolean
-) {
-  if (__DEV__ && !(proxyObject as any)?.[LISTENERS]) {
-    console.warn('Please use proxy object')
-  }
-  let promise: Promise<void> | undefined
-  const ops: Op[] = []
-  const listener: Listener = (op) => {
-    ops.push(op)
-    if (notifyInSync) {
-      callback(ops.splice(0))
-      return
+  subscribe = <T extends object>(
+    proxyObject: T,
+    callback: (ops: Op[]) => void,
+    notifyInSync?: boolean
+  ) => {
+    if (__DEV__ && !(proxyObject as any)?.[LISTENERS]) {
+      console.warn('Please use proxy object')
     }
-    if (!promise) {
-      promise = Promise.resolve().then(() => {
-        promise = undefined
+    let promise: Promise<void> | undefined
+    const ops: Op[] = []
+    const listener: Listener = (op) => {
+      ops.push(op)
+      if (notifyInSync) {
         callback(ops.splice(0))
-      })
+        return
+      }
+      if (!promise) {
+        promise = Promise.resolve().then(() => {
+          promise = undefined
+          callback(ops.splice(0))
+        })
+      }
     }
-  }
-  ;(proxyObject as any)[LISTENERS].add(listener)
-  return () => {
-    ;(proxyObject as any)[LISTENERS].delete(listener)
-  }
-}
-
-type AnyFunction = (...args: any[]) => any
-
-/**
- * This is not a public API.
- * It can be changed without any notice.
- */
-export type INTERNAL_Snapshot<T> = T extends AnyFunction
-  ? T
-  : T extends AsRef
-  ? T
-  : T extends Promise<infer V>
-  ? INTERNAL_Snapshot<V>
-  : {
-      readonly [K in keyof T]: INTERNAL_Snapshot<T[K]>
+    ;(proxyObject as any)[LISTENERS].add(listener)
+    return () => {
+      ;(proxyObject as any)[LISTENERS].delete(listener)
     }
+  },
 
-export function snapshot<T extends object>(
-  proxyObject: T
-): INTERNAL_Snapshot<T> {
-  if (__DEV__ && !(proxyObject as any)?.[SNAPSHOT]) {
-    console.warn('Please use proxy object')
+  snapshot = <T extends object>(proxyObject: T): INTERNAL_Snapshot<T> => {
+    if (__DEV__ && !(proxyObject as any)?.[SNAPSHOT]) {
+      console.warn('Please use proxy object')
+    }
+    return (proxyObject as any)[SNAPSHOT]
   }
-  return (proxyObject as any)[SNAPSHOT]
-}
+) =>
+  [
+    objectIs,
+    refSet,
+    canProxy,
+    ref,
+    VERSION,
+    LISTENERS,
+    SNAPSHOT,
+    PROMISE_RESULT,
+    PROMISE_ERROR,
+    proxyCache,
+    snapshotCache,
+    createSnapshot,
+    globalVersionHolder,
+    proxy,
+    getVersion,
+    subscribe,
+    snapshot,
+  ] as const
 
-export function getHandler<T extends object>(proxyObject: T) {
-  if (__DEV__ && !(proxyObject as any)?.[HANDLER]) {
-    console.warn('Please use proxy object')
-  }
-  return (proxyObject as any)[HANDLER]
-}
+export const [
+  ,
+  ,
+  ,
+  ref,
+  ,
+  ,
+  ,
+  ,
+  ,
+  ,
+  ,
+  ,
+  ,
+  proxy,
+  getVersion,
+  subscribe,
+  snapshot,
+] = buildFunctions()
```

Comment:
I know what you mean, but I'm fearful about the bundle size increase.

This will be second iteration. combining symbols or so will improve a little.

Reply:
Hmm, maybe we don't need to follow the order.

Reply:
Ah, didn't know returning as array vs object resulted in difference in tree-shaking. 👀

Reply:
It's minification, not tree-shaking.
In terms of tree-shaing, this PR makes it really worse, because, for example, `ref` is rarely used and can be tree-shaked previously, but this PR prevents it.
That said, it's fairly micro optimization, and capability would be more important.

Reply:
Wait... I think we still should create public functions outside. Otherwise, useSnapshot must be re-created.